lxd/instance/drivers/driver_lxc.go

```diff
@@ -739,11 +739,16 @@ func (d *lxc) initLXC(config bool) error {
 		}
 	}
 
-	if d.state.OS.CoreScheduling {
+	if d.state.OS.ContainerCoreScheduling {
```

Comment:
If the shared library supports a feature directly we usually never go through LXD ad-hoc APIs. For core scheduling the shared library is aware that core scheduling is used and internally tracks that cookie and automatically applies it during exec/attach as well. An yes, it should be faster as well because running hooks means spawning a subprocess in LXC as well.

Reply:
OK thanks. So should we log a warning if using the LXD approach as its not providing as thorough protection (the exec/attach scenario you mentioned)?

Reply:
I just don't see why we wouldn't try and enable core scheduling even if the shared library doesn't support it as it's not a lot of code to carry.

Reply:
Yes that make sense absolutely. 

I was just thinking that from a support perspective it would be good to know if the container is going to get the "full" support or the "partial" support, seeing as this is a security feature. 

I think it would also be useful for future readers of the code to document the difference between `CoreScheduling` and `ContainerCoreScheduling` so that users can discover what their systems supports and the implications of that.

So perhaps adding a comment to the definition of those variables, and an update in https://github.com/lxc/lxd/blob/master/doc/security.md about Core Scheduling?

Reply:
> OK thanks. So should we log a warning if using the LXD approach as its not providing as thorough protection (the exec/attach scenario you mentioned)?

Hm, I guess we could mention that. Where did you have in mind?

Reply:
Ok, made it work for exec as well.

Reply:
> @brauner Nice! So now the difference between ContainerCoreScheduling and just CoreScheduling for containers is only performance (which is only slight)?

I wouldn't call it "slight" given that in order to make this work we need to constantly spawn additional helper processes either in our subprocesses (forkexec) or in the hooks in LXC (which are also subprocesses already). So with full library support you get rid of a bunch of additional work both for LXD and for the kernel. I thought my previous comment already highlighted that.

And for both qemu and LXC we shouldn't do it behind their backs if they expose an API for this which is how that should be done.